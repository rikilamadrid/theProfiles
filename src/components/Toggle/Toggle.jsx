import React, { useState, useEffect } from 'react';

/** @jsxImportSource @emotion/react */
import { css, jsx } from '@emotion/react';

const Toggle = ({ 'data-testid': dataTestId, label, onChange, checked }) => {
  const [checkedStatus, setCheckedStatus] = useState(checked);

  useEffect(() => {
    setCheckedStatus(checked);
  }, [checked]);

  const changeHandle = (ev) => {
    ev.stopPropagation();
    setCheckedStatus(!checkedStatus);

    if (onChange) {
      onChange(ev, !checkedStatus);
    }
  };

  const stLabel = css`
    display: flex;
    align-items: center;
    width: auto;
    margin-right: 0.5rem;
  `;

  const stToggle = css`
    display: inline-block;
    position: relative;
    cursor: pointer;
    height: 24px;
    width: '36px';

    input {
      display: none;
    }

    .toggle {
      position: relative;
      display: inline-block;
      width: 36px;
      height: 20px;
      background-color: #ededed;
      border-radius: 20px;
      transition: all 0.1s ease-out;

      &:after {
        content: '';
        position: absolute;
        width: 16px;
        height: 16px;
        border-radius: 50%;
        background-color: white;
        top: 2px;
        left: 2px;
        transition: all 0.1s ease-out;
      }
    }

    input:checked ~ .toggle::after {
      left: 18px;
    }

    input:checked ~ .toggle {
      background-color: #1927f0;
      border-color: #163574;
    }
  `;

  return (
    <label css={[stToggle, stLabel]}>
      <input
        data-testid={dataTestId}
        type="checkbox"
        checked={!!checkedStatus}
        onChange={changeHandle}
      />
      {label && <span css={stLabel}>{label}</span>}
      <span className="toggle"></span>
    </label>
  );
};

export default Toggle;
