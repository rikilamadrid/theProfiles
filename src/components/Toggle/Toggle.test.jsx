import React from 'react';

import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import rerender from 'react-test-renderer';

import Toggle from './Toggle';

describe('Toggle component render', () => {
  test('correctly without crashing, default', () => {
    render(<Toggle />);
  });

  test('correctly without crashing, matching snapshot', () => {
    const component = rerender.create(<Toggle />);
    expect(component).toMatchSnapshot();
  });
});

describe('Toggle component property', () => {
  describe('checked', () => {
    test('set to true', () => {
      const { getByTestId } = render(<Toggle data-testid="toggle" checked />);
      expect(getByTestId('toggle')).toBeChecked();
    });

    test('set to false', () => {
      const { getByTestId } = render(<Toggle data-testid="toggle" />);
      expect(getByTestId('toggle')).not.toBeChecked();
    });
  });
});

describe('Toggle component interaction', () => {
  test('check', () => {
    const checkCB = jest.fn();
    const { getByTestId } = render(<Toggle data-testid="toggle" onChange={checkCB} />);
    userEvent.click(getByTestId('toggle'));
    expect(checkCB).toHaveBeenCalled();
  });
});