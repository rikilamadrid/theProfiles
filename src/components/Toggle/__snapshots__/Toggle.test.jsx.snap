// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle component render correctly without crashing, matching snapshot 1`] = `
<label
  className="css-iin6n3-Toggle"
>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="toggle"
  />
</label>
`;
